### ISSUE TITLE
[Question] Performance advantages over powerlevel10k

### ISSUE THREAD

**Issue by sinetoami** [robobenklein/p10k#11]
First of all, sorry for open this issue here when the README tell us for open on GitLab. And sorry my bad english.

I've seeing lots of discution about the powerlevel10k prompt and how great is the improvements working on it and how much great is the `gitstatus` on it too. I use it and i really like it actually. In some of these discutions the Zinc prompt appears as a great prompt using gitstatus too. I saw the zinc seems more simple than powerlevel10k. But in terms of performance, have zinc some advantages over powerlevel10k? Which are these advantages? Zinc really uses `gitstatus`? Can I make my own customizations using the zinc code as base to do this?

For now, that's it. Thanks for your attention!

**Comment by robobenklein** [comment]
No problem, issues on GitLab is mostly a technical thing, since it's easier to track for me there. Discussions here are no problem.

Powerlevel10k was designed to be a drop-in replacement for P9K so it's still using mostly the same configuration model, but in my opinion that model is a little less efficient (along with the amount of registrations and hooks that are created... ex. icons) and the P9K codebase seems a big too large and unwieldy for newcomers to get involved with or understand quickly.

But the biggest point for Zinc over anything like Powerlevel9/10k is that Zinc is designed to be easily extensible as a framework with the default segments being completely optional to load.

Ideally, Zinc segments should be completely portable, i.e. you can import a single file with some definitions and source it, or just write a few of your own functions to create any segment you might want, async or synchronous.

I'm still working on covering all the use cases that people desire in terms of the segment API, but since it's recently gotten more popular it's really helped me see how others desire to use such a system, so it should be getting even better quickly!

**Comment by sinetoami** [comment]
Thanks for your reply!

I had noticed that zinc have a very friedly API which allow us to plugin our own segments. Seems like the p9k/10k custom segments... but with the transparency if it's async or sync. The p9k/10k does not specify clearly, don't have this transparency. For someone who havent skills to programming saphisticated things in zsh(sync or async), it's very painful try to plugin something like language version(like spaceship or alien-minimal does) and tells the prompt to put the information in any side what I want to.

I saw zinc is fast as pure too, which sounds very nice to me because the powerlevel10k it's not so fast as pure( **to me** ) in loading when I open a new prompt for example. But pure loses when the point is git infos without necessarily inside a big git repository. A lot of times pure have a big delay to report to me dirty and upstream informations... with powerlevel10k I havent these problems. Because of that I asked if the zinc uses `gitstatus` in background like powerlevel10k. If I think gives a try to zinc, I want to be sure that I will have the similar performance at this point too. 

Another thing that I actually don't like in powerline10k it's the amount of default segments predefined. It's a lot for some one who just want to use a simple prompt. I think it's better just offer a simple naked prompt and gives us a "how to" wiki section with a sort of segments to plugin by ourself (like fzf does with the "examples" section in the wiki) - *Do you want this segment? Do this... Do you want this another? Do that* , and so on. Can you understand me? Maybe that's your idea about framework if I understand you well and i really apreciate this idea if so.

**Comment by robobenklein** [comment]
I also liked the idea of having barely any default segments provided, but in order to gain traction with users I put what I believe to be the bare minimum to show that the prompt is completely usable out of box.

In order to get people onboard and help demonstrate more of the possible configurations I've made a `presets` folder where all the presets are available to be loaded with the `prompt_zinc_setup` command.

Custom segment support also wasn't quite as advanced as I wanted it to be, when I made Zinc it was basically "run this command and display it's output" without any option to add async to it or do much real work without a subshell.

For initial load times, that's the primary reason why I've kept the code base as small as still readable and also added the `prompt_zinc_compile` helper function to make sure people are using ZSH word code to speed up the parsing of the files.

Gitstatus support seems mature at this point to me, since it should automatically pick up on and use gitstatus when it's loaded before Zinc. (And it can be loaded asynchronously as well while using async vcs_info until the gitstatus backend is ready, see my own configs repo for an example.)

If you want to load Zinc without any of the Zinc segments, just adding the zinc_functions to the fpath and loading it with ZSH's `prompt` should work, although you'll need to make sure you have a full configuration of your own then.

Hopefully that answers all those questions!
